packages/ra-ui-materialui/src/list/datagrid/DatagridBody.tsx

```diff
@@ -45,7 +45,7 @@ const DatagridBody: FC<DatagridBodyProps> = React.forwardRef(
                         hasBulkActions: hasBulkActions && !!selectedIds,
                         hover,
                         id: record.id,
-                        key: record.id,
+                        key: `row${rowIndex}`,
```

Comment:
I think that isn't a good idea for the 90% of the cases where we actually render records. You should only fall back to rowIndex if the record id is undefined.